File: ts/factory/gun/gunFactory.ts
import { Ak47 } from "./gunTypes/ak47";
import { Glob } from "./gunTypes/glob";

export const GunFactory = (gunType: string) => {
  switch (gunType) {
    case 'ak47':
      return new Ak47('ak47', 100);
    case 'glob':
      return new Glob('glob', 50);
    default:
      throw new Error('Invalid gun type');
  }
}

File: go/factory/gun/gunFactory.go
package gun

import "fmt"

func GunFactory(gunType string) (igun, error) {
	if gunType == "glob" {
		return newGlub(), nil
	} else if gunType == "ak47" {
		return newAk47(), nil
	} else {
		return nil, fmt.Errorf("gun type %s is not supported", gunType)
	}

}
